File: ios/IPTVCar/Core/Security/Redaction.swift
import Foundation

/// Masks credential-bearing parts of a URL before it is ever logged.
/// Mirrors android/core/.../security/Redaction.kt — keep both in sync.
enum Redaction {

    private static let sensitiveQueryKeys: Set<String> = ["username", "password", "token", "pass", "user"]

    static func redactQuery(_ urlString: String) -> String {
        guard var components = URLComponents(string: urlString), let items = components.queryItems else {
            return urlString
        }
        components.queryItems = items.map { item in
            if sensitiveQueryKeys.contains(item.name.lowercased()) {
                return URLQueryItem(name: item.name, value: "***")
            }
            return item
        }
        return components.string ?? urlString
    }

    /// Strips Xtream's path-embedded credentials: /live/USER/PASS/id.ext
    static func redactPathCredentials(_ urlString: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: "/(live|movie|series)/[^/]+/[^/]+/") else {
            return urlString
        }
        let range = NSRange(urlString.startIndex..., in: urlString)
        return regex.stringByReplacingMatches(in: urlString, range: range, withTemplate: "/$1/***/***/")
    }

    static func redact(_ urlString: String) -> String {
        redactPathCredentials(redactQuery(urlString))
    }
}

File: android/core/src/main/kotlin/com/iptvcar/core/security/Redaction.kt
package com.iptvcar.core.security

private val SENSITIVE_QUERY_KEYS = listOf("username", "password", "token", "pass", "user")

/**
 * Masks credential-bearing query parameters before a URL is ever logged.
 * Never log a provider/stream URL without passing it through this first.
 */
fun redactUrl(url: String): String {
    val queryStart = url.indexOf('?')
    if (queryStart == -1) return url
    val base = url.substring(0, queryStart)
    val query = url.substring(queryStart + 1)
    val redactedQuery = query.split('&').joinToString("&") { pair ->
        val eq = pair.indexOf('=')
        if (eq == -1) return@joinToString pair
        val key = pair.substring(0, eq)
        if (SENSITIVE_QUERY_KEYS.any { it.equals(key, ignoreCase = true) }) {
            "$key=***"
        } else {
            pair
        }
    }
    return "$base?$redactedQuery"
}

/** Also strips Xtream's path-embedded credentials: /live/USER/PASS/id.ext */
fun redactPathCredentials(url: String): String {
    val pathCredentialRegex = Regex("/(live|movie|series)/[^/]+/[^/]+/")
    return pathCredentialRegex.replace(url) { m -> "/${m.groupValues[1]}/***/***/ " }.trim()
}

fun redact(url: String): String = redactPathCredentials(redactUrl(url))
